# Working with Triangles
## Getting Started
Welcome! We drafted these tutorials to help you get familiar with some of the common functionalities that most actuaries can use in their day-to-day responsibilities. The package also comes with pre-installed datasets that you can play with, which are also used in the tutorials here.

The tutorials assume that you have the basic understanding of commonly used actuarial terms, and can independently perform an actuarial analysis in another tool, such as Microsoft Excel or another actuarial software. Furthermore, it is assumed that you already have some familiarity with Python, and that you have the basic knowledge and experience in using some common packages that are popular in the Python community, such as `pandas` and `numpy`.

This tutorial is linted using [black](https://github.com/psf/black) via [nb_black](https://github.com/dnanhkhoa/nb_black). This step is optional.

In [1]:
%load_ext lab_black

All tutorials and exercises rely on chainladder v0.8.12 and later. It is highly recomended that you keep your packages up-to-date.

For more info on how to update your pakages, visit [Keeping Packages Updated](https://chainladder-python.readthedocs.io/en/latest/library/install.html#keeping-packages-updated).

In [2]:
import pandas as pd
import numpy as np
import chainladder as cl

print("pandas: " + pd.__version__)
print("numpy: " + np.__version__)
print("chainladder: " + cl.__version__)

pandas: 1.4.2
numpy: 1.22.4
chainladder: 0.8.12


Since we will be plotting for quite a bit, here's a magic function in IPython, which sets the backend of matplotlib to the 'inline' backend. With this backend, the output of plotting commands is displayed inline within frontends like the Jupyter notebook, directly below the code cell that produced it. The resulting plots will then also be stored in the notebook document.

In [3]:
%matplotlib inline

## Disclaimer
Note that a lot of the examples shown might not be applicable in a real world scenario, and is only meant to demonstrate some of the functionalities included in the package. The user should always follow all applicable laws, the Code of Professional Conduct, applicable Actuarial Standards of Practice, and exercise their best actuarial judgement.

## Working with a Triangle
Let's begin by looking at an unprocessed triangle data and load it into a `pandas.DataFrame`. We'll use the data `raa`, which is available from the repository.

In [4]:
raa_df = pd.read_csv(
    "https://raw.githubusercontent.com/casact/chainladder-python/master/chainladder/utils/data/raa.csv"
)
raa_df.head(20)

development  origin   values
0          1981    1981   5012.0
1          1982    1982    106.0
2          1983    1983   3410.0
3          1984    1984   5655.0
4          1985    1985   1092.0
5          1986    1986   1513.0
6          1987    1987    557.0
7          1988    1988   1351.0
8          1989    1989   3133.0
9          1990    1990   2063.0
10         1982    1981   8269.0
11         1983    1982   4285.0
12         1984    1983   8992.0
13         1985    1984  11555.0
14         1986    1985   9565.0
15         1987    1986   6445.0
16         1988    1987   4020.0
17         1989    1988   6947.0
18         1990    1989   5395.0
19         1983    1981  10907.0

The data has three columns: 
* development: or valuation time, in this case, valuation year
* origin: or accident date, in this case, accident year
* values: the values recorded for the specific accident date at the specific valuation time (such as incurred losses, paid losses, or claim counts)

A table of loss experience showing total losses for a certain period (origin) at various, regular valuation dates (development), reflect the change in amounts as claims mature and emerge. Older periods in the table will have one more entry than the next youngest period, leading to the triangle shape of the data in the table or any other measure that matures over time from an origin date. Loss triangles can be used to determine loss development for a given risk.

Let's put our data into the triangle format.

In [5]:
raa = cl.Triangle(
    raa_df,
    origin="origin",
    development="development",
    columns="values",
    cumulative=True,
)
raa

12       24       36       48       60       72       84       96       108      120
1981  5012.0   8269.0  10907.0  11805.0  13539.0  16181.0  18009.0  18608.0  18662.0  18834.0
1982   106.0   4285.0   5396.0  10666.0  13782.0  15599.0  15496.0  16169.0  16704.0      NaN
1983  3410.0   8992.0  13873.0  16141.0  18735.0  22214.0  22863.0  23466.0      NaN      NaN
1984  5655.0  11555.0  15766.0  21266.0  23425.0  26083.0  27067.0      NaN      NaN      NaN
1985  1092.0   9565.0  15836.0  22169.0  25955.0  26180.0      NaN      NaN      NaN      NaN
1986  1513.0   6445.0  11702.0  12935.0  15852.0      NaN      NaN      NaN      NaN      NaN
1987   557.0   4020.0  10946.0  12314.0      NaN      NaN      NaN      NaN      NaN      NaN
1988  1351.0   6947.0  13112.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1989  3133.0   5395.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1990  2063.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN

You can also load the example data directly, using `load_sample`:

In [6]:
raa = cl.load_sample("raa")
raa

12       24       36       48       60       72       84       96       108      120
1981  5012.0   8269.0  10907.0  11805.0  13539.0  16181.0  18009.0  18608.0  18662.0  18834.0
1982   106.0   4285.0   5396.0  10666.0  13782.0  15599.0  15496.0  16169.0  16704.0      NaN
1983  3410.0   8992.0  13873.0  16141.0  18735.0  22214.0  22863.0  23466.0      NaN      NaN
1984  5655.0  11555.0  15766.0  21266.0  23425.0  26083.0  27067.0      NaN      NaN      NaN
1985  1092.0   9565.0  15836.0  22169.0  25955.0  26180.0      NaN      NaN      NaN      NaN
1986  1513.0   6445.0  11702.0  12935.0  15852.0      NaN      NaN      NaN      NaN      NaN
1987   557.0   4020.0  10946.0  12314.0      NaN      NaN      NaN      NaN      NaN      NaN
1988  1351.0   6947.0  13112.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1989  3133.0   5395.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1990  2063.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN

A triangle has more properties than just what is displayed. For example we can see the underlying `link_ratio`s, which represent the multiplicative change in amounts from one development period to the next.

In [7]:
raa.link_ratio

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
1981   1.649840  1.319023  1.082332  1.146887  1.195140  1.112972  1.033261  1.002902  1.009217
1982  40.424528  1.259277  1.976649  1.292143  1.131839  0.993397  1.043431  1.033088       NaN
1983   2.636950  1.542816  1.163483  1.160709  1.185695  1.029216  1.026374       NaN       NaN
1984   2.043324  1.364431  1.348852  1.101524  1.113469  1.037726       NaN       NaN       NaN
1985   8.759158  1.655619  1.399912  1.170779  1.008669       NaN       NaN       NaN       NaN
1986   4.259749  1.815671  1.105367  1.225512       NaN       NaN       NaN       NaN       NaN
1987   7.217235  2.722886  1.124977       NaN       NaN       NaN       NaN       NaN       NaN
1988   5.142117  1.887433       NaN       NaN       NaN       NaN       NaN       NaN       NaN
1989   1.721992       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN

We can also view (and manipulate) the `latest_diagonal` of the triangle.

In [8]:
raa.latest_diagonal

1990
1981  18834.0
1982  16704.0
1983  23466.0
1984  27067.0
1985  26180.0
1986  15852.0
1987  12314.0
1988  13112.0
1989   5395.0
1990   2063.0

In [9]:
raa.latest_diagonal / 1000

1990
1981  18.834
1982  16.704
1983  23.466
1984  27.067
1985  26.180
1986  15.852
1987  12.314
1988  13.112
1989   5.395
1990   2.063

The latest diagonal also corresponds to a `valuation_date`. Note that 'valuation_date' is a datetime that is at the terminal timestamp of the period (i.e. the last split second of the year).

In [10]:
raa.valuation_date

Timestamp('1990-12-31 23:59:59.999999999')

We can also tell whether our triangle:
* `is_cumulative`: returns True if the data across the development periods is cumulative, or False if it is incremental.
* `is_ultimate`: returns True if the ultimate values are contained in the triangle.
* `is_val_tri`: returns True if the development period is stated as a valuation data as opposed to an age, i.e. Schedule P style triangle (True) or the more commonly used triangle by development age (False).
* `is_full`: returns True if the triangle has been "squared".

In [11]:
print("Is triangle cumulative?", raa.is_cumulative)
print("Does triangle contain ultimate projections?", raa.is_ultimate)
print("Is this a valuation triangle?", raa.is_val_tri)
print('Has the triangle been "squared"?', raa.is_full)

Is triangle cumulative? True
Does triangle contain ultimate projections? False
Is this a valuation triangle? False
Has the triangle been "squared"? False


We can also inspect the triangle to understand its data granularity with `origin_grain` and `development_grain`.

In [12]:
print("Origin grain:", raa.origin_grain)
print("Development grain:", raa.development_grain)

Origin grain: Y
Development grain: Y


The package supports monthly ("M"), quarterly ("Q"), semester (or semi-annually), ("S") and yearly ("Y") grains for both `origin_grain` and `development_grain`.

## The Triangle Structure
The triangle described so far is a two-dimensional (accident date by valuation date) structure that spans multiple cells of data. This is a useful structure for exploring individual triangles, but becomes more problematic when working with **sets** of triangles. `Pandas` does not have a triangle `dtype`, but if it did, working with sets of triangles would be much more convenient. To facilitate working with more than one triangle at a time the `chainladder.Triangle` acts like a pandas dataframe (with an index and columns) where each cell (row x col) is an individual triangle.  This structure manifests itself as a four-dimensional space. Let's take a look at another dataset `clrd`.

In [13]:
clrd_df = pd.read_csv(
    "https://raw.githubusercontent.com/casact/chainladder-python/master/chainladder/utils/data/clrd.csv"
)
clrd_df

GRCODE                            GRNAME  AccidentYear  \
0          86               Allstate Ins Co Grp          1988   
1          86               Allstate Ins Co Grp          1988   
2          86               Allstate Ins Co Grp          1988   
3          86               Allstate Ins Co Grp          1988   
4          86               Allstate Ins Co Grp          1988   
...       ...                               ...           ...   
42840   44598  College Liability Ins Co Ltd RRG          1995   
42841   44598  College Liability Ins Co Ltd RRG          1995   
42842   44598  College Liability Ins Co Ltd RRG          1996   
42843   44598  College Liability Ins Co Ltd RRG          1996   
42844   44598  College Liability Ins Co Ltd RRG          1997   

       DevelopmentYear  DevelopmentLag  IncurLoss  CumPaidLoss  BulkLoss  \
0                 1988               1     367404        70571    127737   
1                 1989               2     362988       155905     60173   
2                 1990               3     347288       220744     27763   
3                 1991               4     330648       251595     15280   
4                 1992               5     354690       274156     27689   
...                ...             ...        ...          ...       ...   
42840             1996               2        343          249        82   
42841             1997               3        839          575       190   
42842             1996               1        125            6        98   
42843             1997               2         95           17        28   
42844             1997               1         83            4        77   

       EarnedPremDIR  EarnedPremCeded  EarnedPremNet  Single  PostedReserve97  \
0             400699             5957         394742       0           281872   
1             400699             5957         394742       0           281872   
2             400699             5957         394742       0           281872   
3             400699             5957         394742       0           281872   
4             400699             5957         394742       0           281872   
...              ...              ...            ...     ...              ...   
42840            397                0            397       1              630   
42841            397                0            397       1              630   
42842            257                0            257       1              630   
42843            257                0            257       1              630   
42844            256                0            256       1              630   

           LOB  
0       wkcomp  
1       wkcomp  
2       wkcomp  
3       wkcomp  
4       wkcomp  
...        ...  
42840  othliab  
42841  othliab  
42842  othliab  
42843  othliab  
42844  othliab  

[42845 rows x 14 columns]

Let's load the data into the sets of triangles.

In [14]:
clrd = cl.Triangle(
    clrd_df,
    origin="AccidentYear",
    development="DevelopmentYear",
    columns=[
        "IncurLoss",
        "CumPaidLoss",
        "BulkLoss",
        "EarnedPremDIR",
        "EarnedPremCeded",
        "EarnedPremNet",
    ],
    index=["GRNAME", "LOB"],
    cumulative=True,
)
clrd

Triangle Summary
Valuation:                                            1997-12
Grain:                                                   OYDY
Shape:                                       (775, 6, 10, 10)
Index:                                          [GRNAME, LOB]
Columns:    [IncurLoss, CumPaidLoss, BulkLoss, EarnedPremD...

Since 4D strucures do not fit nicely on 2D screens, we see a summary view instead that describes the structure rather than the underlying data itself. 

We see 5 rows of information:
* Valuation: the valuation date.
* Grain: the granularity of the data, O stands for origin, and D stands for development, `OYDY` represents triangles with accident year by development year.
* Shape: contains 4 numbers, represents the 4-D structure. This sample triangle represents a collection of 775x6 or 4,650 triangles that are themselves 10 accident years by 10 development periods.
    * 775: the number of segments, which is the combination of `index`, that represents the data segments. In this case, it is each of the `GRNAME` and `LOB` unique combination.
    * 6: the number of triangles for each segment, which is also the columns `[IncurLoss, CumPaidLoss, BulkLoss, EarnedPremDIR, EarnedPremCeded, EarnedPremNet]`. They could be paid amounts, incurred amounts, reported counts, loss ratios, closure rates, excess losses, premium, etc.
    * 10: the number of accident periods.
    * 10: the number of valuation periods.
* Index: the segmentation level of the triangles.
* Columns: the value types recorded in the triangles.
    
To summarize the set of triangles:
* We have a total of 775 segments, which are at the `GRNAME` and `LOB` level.
* Each segment contains 6 triangles, which are `IncurLoss`, `CumPaidLoss`, `BulkLoss`, `EarnedPremDIR`, `EarnedPremCeded`, `EarnedPremNet`.
* Each triangle is 10 accident years x 10 development periods.

Using `index.head()` allows us to see the first 5 segments in the set of triangles. Note that as the data is loaded, the triangle are sorted by `index`, in this case, `GRNAME` first, then by `LOB`.

In [15]:
clrd.index.head()

GRNAME      LOB
0  Adriatic Ins Co  othliab
1  Adriatic Ins Co   ppauto
2        Aegis Grp  comauto
3        Aegis Grp  othliab
4        Aegis Grp   ppauto

Under the hood, the data structure is a `numpy.ndarray` with the equivalent shape.  Like pandas, you can directly access the underlying numpy structure with the `values` property. By exposing the underlying `ndarray` you are free to manipulate the underlying data directly with numpy should that be an easier route to solving a problem.

In [16]:
print("Data structure of clrd:", type(clrd.values))
print("Sum of all data values:", np.nansum(clrd.values))

Data structure of clrd: <class 'numpy.ndarray'>
Sum of all data values: 3661713596.0


Keep in mind though, the `chainladder.Triangle` has several methods and properties beyond the raw numpy representation and these are kept in sync by using the `chainladder.Triangle` directly.

## Pandas-Style Slicing
As mentioned, the 4D structure is intended to behave like a pandas `DataFrame`.  Like pandas, we can subset a dataframe by referencing individual columns by name.

In [17]:
clrd[["CumPaidLoss", "IncurLoss", "BulkLoss"]]

Triangle Summary
Valuation:                             1997-12
Grain:                                    OYDY
Shape:                        (775, 3, 10, 10)
Index:                           [GRNAME, LOB]
Columns:    [CumPaidLoss, IncurLoss, BulkLoss]

We can also boolean-index the rows of the Triangle.

In [18]:
clrd[clrd["LOB"] == "wkcomp"]

Triangle Summary
Valuation:                                            1997-12
Grain:                                                   OYDY
Shape:                                       (132, 6, 10, 10)
Index:                                          [GRNAME, LOB]
Columns:    [IncurLoss, CumPaidLoss, BulkLoss, EarnedPremD...

We can even use the typical `loc`, `iloc` functionality similar to `pandas` to access subsets of data.  These features can be chained together as much as you want.

In [19]:
clrd.loc["Allstate Ins Co Grp"].iloc[-1]["CumPaidLoss"]

12        24        36        48        60        72        84        96        108       120
1988  70571.0  155905.0  220744.0  251595.0  274156.0  287676.0  298499.0  304873.0  321808.0  325322.0
1989  66547.0  136447.0  179142.0  211343.0  231430.0  244750.0  254557.0  270059.0  273873.0       NaN
1990  52233.0  133370.0  178444.0  204442.0  222193.0  232940.0  253337.0  256788.0       NaN       NaN
1991  59315.0  128051.0  169793.0  196685.0  213165.0  234676.0  239195.0       NaN       NaN       NaN
1992  39991.0   89873.0  114117.0  133003.0  154362.0  159496.0       NaN       NaN       NaN       NaN
1993  19744.0   47229.0   61909.0   85099.0   87215.0       NaN       NaN       NaN       NaN       NaN
1994  20379.0   46773.0   88636.0   91077.0       NaN       NaN       NaN       NaN       NaN       NaN
1995  18756.0   84712.0   87311.0       NaN       NaN       NaN       NaN       NaN       NaN       NaN
1996  42609.0   44916.0       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN
1997    691.0       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN

## Pandas-Style Arithmetic
With complete flexibility in the ability to slice subsets of triangles, we can use basic arithmetic to derive new triangles, which is commonly used as diagnostics to explore trends.

In [20]:
clrd["CaseIncurLoss"] = clrd["IncurLoss"] - clrd["BulkLoss"]
clrd["PaidToInc"] = clrd["CumPaidLoss"] / clrd["CaseIncurLoss"]
clrd[["CaseIncurLoss", "PaidToInc"]]

Triangle Summary
Valuation:                     1997-12
Grain:                            OYDY
Shape:                (775, 2, 10, 10)
Index:                   [GRNAME, LOB]
Columns:    [CaseIncurLoss, PaidToInc]

We can also aggregating the values across all triangles into one triangle.

In [21]:
clrd["CumPaidLoss"].sum()

12          24          36          48          60          72          84          96          108         120
1988  3577780.0   7059966.0   8826151.0   9862687.0  10474698.0  10814576.0  10994014.0  11091363.0  11171590.0  11203949.0
1989  4090680.0   7964702.0   9937520.0  11098588.0  11766488.0  12118790.0  12311629.0  12434826.0  12492899.0         NaN
1990  4578442.0   8808486.0  10985347.0  12229001.0  12878545.0  13238667.0  13452993.0  13559557.0         NaN         NaN
1991  4648756.0   8961755.0  11154244.0  12409592.0  13092037.0  13447481.0  13642414.0         NaN         NaN         NaN
1992  5139142.0   9757699.0  12027983.0  13289485.0  13992821.0  14347271.0         NaN         NaN         NaN         NaN
1993  5653379.0  10599423.0  12953812.0  14292516.0  15005138.0         NaN         NaN         NaN         NaN         NaN
1994  6246447.0  11394960.0  13845764.0  15249326.0         NaN         NaN         NaN         NaN         NaN         NaN
1995  6473843.0  11612151.0  14010098.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN
1996  6591599.0  11473912.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN
1997  6451896.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN

We can also construct a paid loss ratio triangle against `EarnedPremNet`.

In [22]:
clrd["CumPaidLoss"].sum() / clrd["EarnedPremNet"].sum()

12        24        36        48        60        72        84        96        108       120
1988  0.258115  0.509334  0.636753  0.711533  0.755686  0.780206  0.793152  0.800175  0.805963  0.808297
1989  0.268634  0.523039  0.652594  0.728841  0.772701  0.795837  0.808501  0.816591  0.820405       NaN
1990  0.272814  0.524868  0.654580  0.728685  0.767389  0.788847  0.801618  0.807968       NaN       NaN
1991  0.253951  0.489561  0.609332  0.677909  0.715189  0.734607  0.745255       NaN       NaN       NaN
1992  0.259518  0.492747  0.607392  0.671096  0.706613  0.724512       NaN       NaN       NaN       NaN
1993  0.264522  0.495948  0.606111  0.668749  0.702092       NaN       NaN       NaN       NaN       NaN
1994  0.270903  0.494190  0.600480  0.661351       NaN       NaN       NaN       NaN       NaN       NaN
1995  0.265095  0.475502  0.573694       NaN       NaN       NaN       NaN       NaN       NaN       NaN
1996  0.263451  0.458585       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN
1997  0.255201       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN

Aggregating all segments together is interesting, but it is often more useful to aggregate across segments using `groupby`.  For example, we may want to group the triangles by line of business and get a sum across all companies for each industry.

In [23]:
clrd.groupby("LOB").sum()

Triangle Summary
Valuation:                                            1997-12
Grain:                                                   OYDY
Shape:                                         (6, 8, 10, 10)
Index:                                                  [LOB]
Columns:    [IncurLoss, CumPaidLoss, BulkLoss, EarnedPremD...

The shape is (**6**, 8, 10, 10) because now we have 6 LOBs with 8 triangles for each LOB. We can also note that `index` is now on `[LOB]` only.

In [24]:
np.unique(clrd["LOB"])

array(['comauto', 'medmal', 'othliab', 'ppauto', 'prodliab', 'wkcomp'],
      dtype=object)

The aggregation functions, e.g. `sum`, `mean`, `std`, `min`, `max`, etc. don't have to just apply to the `index` axis.  You can apply them to any of the four axes in the triangle object, which are `segments` (axis `0`, or the index axis), `columns` (axis `1`, the various financial fields), `origin` (axis `2`, across triangle rows), and `development` period (axis `3`, across triangle columns). You can also use either the axis name or number. Let's try to sum all of the segments (`[GRNAME, LOB]`) and columns (`[IncurLoss, CumPaidLoss, BulkLoss, EarnedPremDIR, EarnedPremCeded, EarnedPremNet]`) using axis name and number, respectively.

In [25]:
clrd.sum(axis="index").sum(axis="segments")

Triangle Summary
Valuation:                                            1997-12
Grain:                                                   OYDY
Shape:                                         (1, 8, 10, 10)
Index:                                          [GRNAME, LOB]
Columns:    [IncurLoss, CumPaidLoss, BulkLoss, EarnedPremD...

In [26]:
clrd.sum(axis=0).sum(axis=1)

12            24            36            48            60            72            84            96            108           120
1988  5.638773e+07  5.992852e+07  6.165346e+07  6.264463e+07  6.318271e+07  6.343808e+07  6.344090e+07  6.345211e+07  6.352298e+07  6.351818e+07
1989  6.284045e+07  6.670557e+07  6.866693e+07  6.970228e+07  7.018801e+07  7.035568e+07  7.038986e+07  7.045316e+07  7.048327e+07           NaN
1990  7.006497e+07  7.408439e+07  7.587928e+07  7.681299e+07  7.717957e+07  7.724032e+07  7.728394e+07  7.734559e+07           NaN           NaN
1991  7.461161e+07  7.851312e+07  8.023591e+07  8.096706e+07  8.117888e+07  8.118548e+07  8.127864e+07           NaN           NaN           NaN
1992  8.121379e+07  8.508937e+07  8.644388e+07  8.678311e+07  8.690861e+07  8.708664e+07           NaN           NaN           NaN           NaN
1993  8.789623e+07  9.168546e+07  9.301849e+07  9.316781e+07  9.347308e+07           NaN           NaN           NaN           NaN           NaN
1994  9.459370e+07  9.813072e+07  9.907178e+07  9.980912e+07           NaN           NaN           NaN           NaN           NaN           NaN
1995  9.772281e+07  1.011923e+08  1.020567e+08           NaN           NaN           NaN           NaN           NaN           NaN           NaN
1996  9.849793e+07  1.009177e+08           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
1997  9.683222e+07           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN

## Accessor Methods

`Pandas` has special "accessor" methods for `str` and `dt`.  These allow for the manipulation of data within each cell of data:

```python
# splits lastname from first name by a comma-delimiter
df['Last_First'].str.split(',')

# pulls the year out of each date in a dataframe column
df['Accident Date'].dt.year 
```

`chainladder` also has special "accessor" methods to help us manipulate the `origin`, `development` and `valuation` vectors of a triangle.

We may want to extract only the latest accident period for every triangle.

In [27]:
clrd[clrd.origin == clrd.origin.max()]

Triangle Summary
Valuation:                                            1997-12
Grain:                                                   OYDY
Shape:                                        (775, 8, 1, 10)
Index:                                          [GRNAME, LOB]
Columns:    [IncurLoss, CumPaidLoss, BulkLoss, EarnedPremD...

Note that this triangle has only 1 row; however, all of the columns would exist, but only the youngest age would have values.

We may want to extract particular diagonals from our triangles using its `valuation` vector.

In [28]:
clrd[(clrd.valuation >= "1994") & (clrd.valuation < "1995")]["CumPaidLoss"].sum()

12          24          36          48          60          72          84
1988        NaN         NaN         NaN         NaN         NaN         NaN  10994014.0
1989        NaN         NaN         NaN         NaN         NaN  12118790.0         NaN
1990        NaN         NaN         NaN         NaN  12878545.0         NaN         NaN
1991        NaN         NaN         NaN  12409592.0         NaN         NaN         NaN
1992        NaN         NaN  12027983.0         NaN         NaN         NaN         NaN
1993        NaN  10599423.0         NaN         NaN         NaN         NaN         NaN
1994  6246447.0         NaN         NaN         NaN         NaN         NaN         NaN

We may even want to slice particular development periods to explore aspects of our data by development age. For example, we can look at the development factors between ages 24 and 36.

In [29]:
clrd[(clrd.development > 12) & (clrd.development <= 36)]["CumPaidLoss"].sum().link_ratio

24-36
1988  1.250169
1989  1.247695
1990  1.247132
1991  1.244650
1992  1.232666
1993  1.222124
1994  1.215078
1995  1.206503
1996       NaN
1997       NaN

## Moving Back to Pandas
When the shape of a `Triangle` object can be expressed as a 2D structure (i.e. two of its four axes have a length of 1) or less, you can use the `to_frame` method to convert your data into a `pandas.DataFrame`. Let's pick only one financial field, `CumPaidLoss` and only the latest diagonal, with `latest_diagonal`. We are now left with LOBs and origin period as our 2 axes.

In [30]:
clrd.groupby("LOB").sum().latest_diagonal["CumPaidLoss"]

Triangle Summary
Valuation:          1997-12
Grain:                 OYDY
Shape:        (6, 1, 10, 1)
Index:                [LOB]
Columns:      [CumPaidLoss]

In [31]:
clrd.groupby("LOB").sum().latest_diagonal["CumPaidLoss"].to_frame(
    origin_as_datetime=True
).astype(int)

origin       1988     1989      1990      1991      1992      1993      1994  \
LOB                                                                            
comauto    626097   674441    718396    711762    731033    762039    768095   
medmal     217239   222707    235717    275923    267007    276235    252449   
othliab    317889   350684    361103    426085    389250    434995    402244   
ppauto    8690036  9823747  10728411  10713621  11555121  12249826  12600432   
prodliab   110973   112614    121255    100276     76059     94462    111264   
wkcomp    1241715  1308706   1394675   1414747   1328801   1187581   1114842   

origin        1995     1996     1997  
LOB                                   
comauto     675166   510191   272342  
medmal      209222   107474    20361  
othliab     294332   191258    54130  
ppauto    11807279  9900842  5754249  
prodliab     62018    28107    10682  
wkcomp      962081   736040   340132

Note that we added `origin_as_datetime=True` inside of `to_frame(...)` as the package is under-going a deprecation cycle. You can also execute the same code without the `origin_as_datetime=...`, but you will get a warning.

In [32]:
clrd.groupby("LOB").sum().latest_diagonal["CumPaidLoss"].to_frame().astype(int)

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/chainladder/core/pandas.py:57: UserWarning: In an upcoming version of the package, `origin_as_datetime` will be defaulted to `True` in to_frame(...), use `origin_as_datetime=False` to preserve current setting.
  warnings.warn(warning)


origin       1988     1989      1990      1991      1992      1993      1994  \
LOB                                                                            
comauto    626097   674441    718396    711762    731033    762039    768095   
medmal     217239   222707    235717    275923    267007    276235    252449   
othliab    317889   350684    361103    426085    389250    434995    402244   
ppauto    8690036  9823747  10728411  10713621  11555121  12249826  12600432   
prodliab   110973   112614    121255    100276     76059     94462    111264   
wkcomp    1241715  1308706   1394675   1414747   1328801   1187581   1114842   

origin        1995     1996     1997  
LOB                                   
comauto     675166   510191   272342  
medmal      209222   107474    20361  
othliab     294332   191258    54130  
ppauto    11807279  9900842  5754249  
prodliab     62018    28107    10682  
wkcomp      962081   736040   340132

We can also aggregate process away 3 dimensions, then use `to_frame`.

In [33]:
clrd[clrd.origin == "1990"].groupby("LOB").sum().latest_diagonal[
    "CumPaidLoss"
].to_frame(origin_as_datetime=True).astype(int)

LOB
comauto       718396
medmal        235717
othliab       361103
ppauto      10728411
prodliab      121255
wkcomp       1394675
dtype: int64

## Exercises
Use the `clrd` dataset for all of the exercises.

1. How do we create a new column named "NetPaidLossRatio" in the triangle using the existing columns?

In [34]:
clrd["NetPaidLossRatio"] = clrd["CumPaidLoss"] / clrd["EarnedPremNet"]

2. What is the highest paid loss ratio for across all segments for origin 1997 at age 12?

In [35]:
clrd[clrd.origin == "1997"][clrd.development == 12]["NetPaidLossRatio"].max()

4.769123134328358

3. How do we subset the overall triangle to just include "Alaska Nat Ins Co"?

In [36]:
clrd[clrd["GRNAME"] == "Alaska Nat Ins Co"]

Triangle Summary
Valuation:                                            1997-12
Grain:                                                   OYDY
Shape:                                         (4, 9, 10, 10)
Index:                                          [GRNAME, LOB]
Columns:    [IncurLoss, CumPaidLoss, BulkLoss, EarnedPremD...

4. How do we create a triangle subset that includes all triangles for companies with names starting with the letter "B"?

In [37]:
clrd[clrd["GRNAME"].str[0] == "B"]

Triangle Summary
Valuation:                                            1997-12
Grain:                                                   OYDY
Shape:                                        (31, 9, 10, 10)
Index:                                          [GRNAME, LOB]
Columns:    [IncurLoss, CumPaidLoss, BulkLoss, EarnedPremD...

5. Which are the top 5 companies by net premium share for in 1990?

In [38]:
clrd.latest_diagonal.groupby("GRNAME").sum()[clrd.origin == "1990"][
    "EarnedPremNet"
].to_frame(origin_as_datetime=True).sort_values(ascending=False).iloc[0:5]

GRNAME
State Farm Mut Grp                    10532675.0
United Services Automobile Asn Grp     1378791.0
Federal Ins Co Grp                      477150.0
New Jersey Manufacturers Grp            383870.0
FL Farm Bureau Grp                      342036.0
dtype: float64

## Initializing a Triangle With Your Own Data
The `chainladder.Triangle`  class is designed to ingest `pandas.DataFrame` objects. However, you do not need to worry about shaping the dataframe into triangle format yourself. This happens at the time you ingest the data.

Let's look at the initialization signature and its docstring.

In [39]:
cl.Triangle?

Init signature:
cl.Triangle(
    data=None,
    origin=None,
    development=None,
    columns=None,
    index=None,
    origin_format=None,
    development_format=None,
    cumulative=None,
    array_backend=None,
    pattern=False,
    trailing=True,
    *args,
    **kwargs,
)
Docstring:     
The core data structure of the chainladder package

Parameters
----------
data: DataFrame
    A single dataframe that contains columns represeting all other
    arguments to the Triangle constructor
origin: str or list
     A representation of the accident, reporting or more generally the
     origin period of the triangle that will map to the Origin dimension
development: str or list
    A representation of the development/valuation periods of the triangle
    that will map to the Development dimension
columns: str or list
    A representation of the numeric data of the triangle that will map to
    the columns dimension.  If None, then a single 'Total' key will be
    generated.
index: str or 

Let's use a new dataset `prism` to construct our triangles.

In [40]:
prism_df = pd.read_csv(
    "https://raw.githubusercontent.com/casact/chainladder-python/master/chainladder/utils/data/prism.csv"
)
prism_df.head()

ClaimNo AccidentDate ReportDate  Line      Type  ClaimLiability     Limit  \
0        1   2008-01-22  4/19/2010  Home  Dwelling           False  300000.0   
1        2   2008-01-02  4/20/2010  Home  Dwelling           False  200000.0   
2        3   2008-01-01  9/23/2009  Home  Dwelling            True  200000.0   
3        4   2008-01-02  7/25/2009  Home  Dwelling            True  200000.0   
4        5   2008-01-16  12/7/2009  Home  Dwelling            True  200000.0   

   Deductible  TotalPayment PaymentDate   CloseDate  Status  reportedCount  \
0       20000       0.00000  2010-10-08   10/8/2010  CLOSED              1   
1       20000       0.00000  2010-11-30  11/30/2010  CLOSED              1   
2       20000  115744.77370  2010-02-17   2/17/2010  CLOSED              1   
3       20000   63678.87713  2009-11-18  11/18/2009  CLOSED              1   
4       20000  112175.55590  2010-04-30   4/30/2010  CLOSED              1   

   closedPaidCount  closedUnPaidCount  openCount          Paid  Outstanding  \
0                0                  1          0       0.00000            0   
1                0                  1          0       0.00000            0   
2                1                  0          0  115744.77370            0   
3                1                  0          0   63678.87713            0   
4                1                  0          0  112175.55590            0   

       Incurred  
0       0.00000  
1       0.00000  
2  115744.77370  
3   63678.87713  
4  112175.55590

In [41]:
prism_df.dtypes

ClaimNo                int64
AccidentDate          object
ReportDate            object
Line                  object
Type                  object
ClaimLiability          bool
Limit                float64
Deductible             int64
TotalPayment         float64
PaymentDate           object
CloseDate             object
Status                object
reportedCount          int64
closedPaidCount        int64
closedUnPaidCount      int64
openCount              int64
Paid                 float64
Outstanding            int64
Incurred             float64
dtype: object

We must specify the `origin`, `development`, `columns`, and `cumulative` to create a triangle object. By limiting our columns to one measure and not specifying an index, we can create a single triangle. For example, if we are only interested in the paid triangle. Because the data we have is transactional data, the `cumulative` variable should be set to `False`.

In [42]:
prism = cl.Triangle(
    data=prism_df,
    origin="AccidentDate",
    development="PaymentDate",
    columns="Paid",
    cumulative=False,
)
prism

1             2              3              4              5              6              7              8              9              10             11             12             13             14             15             16             17             18             19             20             21             22             23             24             25             26            27            28            29            30            31            32            33            34            35            36            37             38             39             40             41            42            43            44            45            46            47            48            49            50            51            52            53            54            55            56            57            58            59            60            61            62            63            64           65            66            67           68            69   70            71            72          73           74       75            76           77   78           79           80   81          82          83      84   85      86            87   88   89           90           91   92   93           94   95   96   97           98   99   100  101  102  103  104  105     106  107  108  109  110  111  112  113  114  115  116  117  118  119  120
2008-01           NaN           NaN   46915.015123   19898.778972   57215.983936   89783.258084   18301.827646   95054.432894   10332.363153   91772.229823   50403.256100   25904.392469   20363.039991   53973.827076   42845.725397   72887.347019   45247.565785   41837.701048   19000.000000   20681.849010    7437.519458   66622.733500  277594.082220   21437.716875   31795.146380  818091.638041  3.775590e+05  6.231213e+05  4.961923e+05  3.729639e+05  1.613041e+06  1.510690e+06  1.463860e+06  2.705437e+06  1.258599e+06  6.154958e+05  3.662470e+05  285689.926240  187000.000000            NaN    8733.822202           NaN   7324.582575           NaN           NaN           NaN           NaN           NaN           NaN   6908.747829           NaN           NaN           NaN           NaN           NaN           NaN           NaN   5659.301494           NaN           NaN   7509.548931           NaN           NaN           NaN          NaN           NaN           NaN          NaN           NaN  NaN           NaN           NaN         NaN          NaN      NaN           NaN          NaN  NaN          NaN          NaN  NaN         NaN         NaN     NaN  NaN     NaN           NaN  NaN  NaN          NaN          NaN  NaN  NaN          NaN  NaN  NaN  NaN          NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2008-02           NaN  28749.365091   22109.176517   79032.630204   63454.689224   59993.109001   64682.696457   68501.669363   33694.528307   71670.163132   10845.932008   76992.306674   45118.695949   31260.402970   31628.854136   44746.068059    5437.649534   28559.099760   20605.070884   23776.507583  243735.616281  127662.219754   63657.136032   23716.146264  341144.987773  963831.511040  1.269554e+06  7.748447e+05  9.630401e+05  1.733778e+06  1.207842e+06  1.042254e+06  1.637606e+06  7.552154e+05  8.835821e+05  8.399512e+05  1.800000e+05  180000.000000  360000.000000            NaN   14000.000000           NaN           NaN           NaN   7000.000000           NaN  13471.002738   9292.483535           NaN   2696.680976           NaN           NaN           NaN   4607.561736  19000.000000           NaN  29395.358080           NaN           NaN           NaN           NaN           NaN           NaN           NaN          NaN           NaN           NaN  5688.953874           NaN  NaN   6803.309901           NaN         NaN          NaN      NaN           NaN          NaN  NaN          NaN          NaN  NaN         NaN         NaN     NaN  NaN     NaN           NaN  NaN  NaN          NaN          NaN  NaN  NaN      

Note that the lowest (most-detailed) grain supported is the monthly grain, so the triangle above is aggregated to the OMDM level.

In [43]:
prism.origin_grain

'M'

In [44]:
prism.development_grain

'M'

If we want to include more columns or indices we can certainly do so. Note that as we do this, we move into the 4D arena changing the display of the overall object.

In [45]:
prism = cl.Triangle(
    data=prism_df,
    origin="AccidentDate",
    development="PaymentDate",
    columns=["Paid", "Incurred"],
    cumulative=False,
)
prism

Triangle Summary
Valuation:           2017-12
Grain:                  OMDM
Shape:      (1, 2, 120, 120)
Index:               [Total]
Columns:    [Paid, Incurred]

`Pandas` has wonderful datetime inference functionality that the `Triangle` heavily uses to infer origin and development granularity. Even still, there are occassions where date format inferences can fail. It is often better to explicitly tell the triangle the date format, and is usually good pratice to explicitly state the date format instead.

In [46]:
prism_df.head()

ClaimNo AccidentDate ReportDate  Line      Type  ClaimLiability     Limit  \
0        1   2008-01-22  4/19/2010  Home  Dwelling           False  300000.0   
1        2   2008-01-02  4/20/2010  Home  Dwelling           False  200000.0   
2        3   2008-01-01  9/23/2009  Home  Dwelling            True  200000.0   
3        4   2008-01-02  7/25/2009  Home  Dwelling            True  200000.0   
4        5   2008-01-16  12/7/2009  Home  Dwelling            True  200000.0   

   Deductible  TotalPayment PaymentDate   CloseDate  Status  reportedCount  \
0       20000       0.00000  2010-10-08   10/8/2010  CLOSED              1   
1       20000       0.00000  2010-11-30  11/30/2010  CLOSED              1   
2       20000  115744.77370  2010-02-17   2/17/2010  CLOSED              1   
3       20000   63678.87713  2009-11-18  11/18/2009  CLOSED              1   
4       20000  112175.55590  2010-04-30   4/30/2010  CLOSED              1   

   closedPaidCount  closedUnPaidCount  openCount          Paid  Outstanding  \
0                0                  1          0       0.00000            0   
1                0                  1          0       0.00000            0   
2                1                  0          0  115744.77370            0   
3                1                  0          0   63678.87713            0   
4                1                  0          0  112175.55590            0   

       Incurred  
0       0.00000  
1       0.00000  
2  115744.77370  
3   63678.87713  
4  112175.55590

In [47]:
prism = cl.Triangle(
    data=prism_df,
    origin="AccidentDate",
    origin_format="%Y-%m-%d",
    development="PaymentDate",
    development_format="%Y-%m-%d",
    columns=["Paid", "Incurred"],
    cumulative=False,
)
prism

Triangle Summary
Valuation:           2017-12
Grain:                  OMDM
Shape:      (1, 2, 120, 120)
Index:               [Total]
Columns:    [Paid, Incurred]

Up until now, we've been playing with symmetric triangles (i.e. `orgin` and `development` periods have the same grain). However, nothing precludes us from having a different grain. Often times in practice the `development` axis is more granular than the `origin` axis. All the functionality available to symmetric triangles works equally well for asymmetric triangles.

In [48]:
prism_df["AccYr"] = prism_df["AccidentDate"].str[:4]

prism = cl.Triangle(
    data=prism_df,
    origin="AccYr",
    origin_format="%Y",
    development="PaymentDate",
    development_format="%Y-%m-%d",
    columns=["Paid", "Incurred"],
    cumulative=False,
)
prism["Paid"]

1             2              3              4              5              6              7              8              9              10             11            12            13            14            15            16            17             18            19             20            21            22             23            24            25            26            27            28            29            30            31            32            33            34            35            36            37            38            39            40            41            42            43            44            45            46            47            48             49             50            51            52            53            54            55            56            57            58            59            60            61            62            63            64            65            66            67            68            69            70            71            72            73            74            75            76           77            78           79           80            81          82      83            84   85   86      87           88           89      90       91          92           93   94   95   96   97   98   99            100  101  102  103  104  105  106  107  108  109  110  111  112  113     114  115  116  117  118  119  120
2008           NaN           NaN   75664.380214   90813.742423  194956.103471  300085.418824  347486.651714  400010.617578  356794.502531  598532.326449  641536.425988  3.983742e+05  7.370244e+05  7.671280e+05  7.341962e+05  5.731743e+05  6.011886e+05  507487.238890  5.262003e+05  454101.038664  4.592687e+05  5.712807e+05  947098.698678  9.086826e+05  3.701474e+05  1.562406e+06  2.415006e+06  2.784466e+06  2.704362e+06  4.546247e+06  4.878159e+06  6.097529e+06  6.037435e+06  1.253314e+07  8.906121e+06  1.058691e+07  7.956213e+06  9.153212e+06  1.077999e+07  1.013129e+07  7.953316e+06  9.281836e+06  5.951154e+06  6.502800e+06  3.651396e+06  2.434386e+06  8.061479e+05  1.128002e+06  317472.619664   26873.221313  75318.053240  54505.399964  40676.115005  10118.603300  11607.561736  23143.087336   7000.000000  42680.332727  30727.022825           NaN  35182.690060  11722.264203  19491.773256   8390.432613           NaN           NaN   1578.971749  29980.600067  34636.956810           NaN   5932.496097  22936.033691           NaN   8351.518321           NaN  21563.293691          NaN  13274.484130  6905.158619  7051.450295           NaN         NaN     NaN  19000.000000  NaN  NaN     NaN          NaN  4751.868062  7000.0  19000.0  5182.60652          NaN  NaN  NaN  NaN  NaN  NaN  NaN  12411.369291  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  7000.0  NaN  NaN  NaN  NaN  NaN  NaN
2009           NaN  24326.814665   80376.045998  136856.648098  175736.898977  297690.337140  358624.070468  325461.635602  529831.969293  540216.723927  556830.668443  5.834332e+05  7.671930e+05  4.859745e+05  7.038561e+05  7.887551e+05  8.759424e+05  714801.340039  5.985041e+05  423752.861179  3.619214e+05  5.596524e+05  634659.052995  4.785295e+05  4.712591e+05  1.647277e+06  1.521688e+06  1.757110e+06  3.142084e+06  4.045939e+06  5.090424e+06  8.524774e+06  8.998482e+06  9.623510e+06  1.323659e+07  1.166428e+07  1.111210e+07  1.004107e+07  1.042377e+07  1.159419e+07  7.894366e+06  6.431454e+06  6.561179e+06  6.370337e+06  2.442785e+06  2.164221e+06  1.056090e+06  1.528802e+05   31965.269691  295714.535436  51422.895883  52724.444105  45404.464826  20357.031655  25607.355442  30396.154054   3462.412866  18329.878692  31737.059878  21549.657455  12265.587530   7254.649009  14735.452863   4555.892746   3911.686172  17247.273140  12347.917460  23092.780648   2222.591893           NaN           NaN           NaN           NaN           NaN           NaN   1656.505479          NaN   2041.114146          NaN          NaN  28594.422270  7000.00000     NaN           N

While exposure triangles make sense for auditable lines such as workers' compensation lines, most other lines of business' exposures can be expressed as a 1D array (along origin period) as exposures do not develop over time. `chainladder` arithmetic requires that operations happen between a triangle and either an `int`, `float`, or another `Triangle`. To create a 1D exposure array, simply omit the `development` argument at initialization.

The `prism` data does not consist of exposure data, but we can contrive one. Let's assume that the premium is thrice the incurred amount.

In [49]:
prism_df["Premium"] = 3 * prism_df["Incurred"]

prism = cl.Triangle(
    data=prism_df,
    origin="AccidentDate",
    origin_format="%Y-%m-%d",
    columns="Premium",
    cumulative=False,
)

prism

2017-12
2008-01  4.187229e+07
2008-02  4.354268e+07
2008-03  3.592080e+07
2008-04  3.052159e+07
2008-05  4.680761e+07
2008-06  3.592624e+07
2008-07  3.330561e+07
2008-08  3.566795e+07
2008-09  3.677317e+07
2008-10  3.685877e+07
2008-11  4.193904e+07
2008-12  3.471688e+07
2009-01  4.043509e+07
2009-02  5.180367e+07
2009-03  3.827396e+07
2009-04  4.404504e+07
2009-05  3.857352e+07
2009-06  3.467500e+07
2009-07  3.937263e+07
2009-08  4.193669e+07
2009-09  3.912008e+07
2009-10  3.348542e+07
2009-11  3.588476e+07
2009-12  3.564036e+07
2010-01  3.918751e+07
2010-02  3.640811e+07
2010-03  4.610033e+07
2010-04  3.931191e+07
2010-05  3.861744e+07
2010-06  3.820465e+07
2010-07  3.901837e+07
2010-08  4.156632e+07
2010-09  3.908993e+07
2010-10  4.236063e+07
2010-11  4.164149e+07
2010-12  4.410378e+07
2011-01  4.455028e+07
2011-02  4.022343e+07
2011-03  4.083951e+07
2011-04  4.290512e+07
2011-05  3.799259e+07
2011-06  4.000331e+07
2011-07  4.249876e+07
2011-08  3.861800e+07
2011-09  4.187191e+07
2011-10  4.184378e+07
2011-11  3.917826e+07
2011-12  3.694711e+07
2012-01  3.788083e+07
2012-02  4.270029e+07
2012-03  4.653762e+07
2012-04  3.785465e+07
2012-05  3.693307e+07
2012-06  3.914957e+07
2012-07  3.893848e+07
2012-08  3.586909e+07
2012-09  3.110809e+07
2012-10  4.205510e+07
2012-11  3.560778e+07
2012-12  3.382385e+07
2013-01  4.813855e+07
2013-02  3.458431e+07
2013-03  4.096955e+07
2013-04  3.508083e+07
2013-05  4.537744e+07
2013-06  4.477441e+07
2013-07  4.143155e+07
2013-08  4.576263e+07
2013-09  4.198754e+07
2013-10  4.120169e+07
2013-11  3.225811e+07
2013-12  3.345250e+07
2014-01  4.036345e+07
2014-02  3.864147e+07
2014-03  3.639684e+07
2014-04  3.578460e+07
2014-05  3.745228e+07
2014-06  4.515285e+07
2014-07  3.963638e+07
2014-08  3.671515e+07
2014-09  4.552311e+07
2014-10  3.117016e+07
2014-11  4.338274e+07
2014-12  3.469077e+07
2015-01  3.697209e+07
2015-02  3.418734e+07
2015-03  3.540646e+07
2015-04  3.213757e+07
2015-05  2.365461e+07
2015-06  1.670796e+07
2015-07  1.807573e+07
2015-08  1.220446e+07
2015-09  1.120974e+07
2015-10  7.647748e+06
2015-11  6.591418e+06
2015-12  7.948603e+06
2016-01  5.502819e+06
2016-02  5.677216e+06
2016-03  4.541502e+06
2016-04  4.938965e+06
2016-05  4.522476e+06
2016-06  4.664779e+06
2016-07  5.134766e+06
2016-08  4.554252e+06
2016-09  4.406520e+06
2016-10  4.278755e+06
2016-11  3.955199e+06
2016-12  3.586924e+06
2017-01  3.446326e+06
2017-02  3.095700e+06
2017-03  2.680826e+06
2017-04  2.634972e+06
2017-05  2.057550e+06
2017-06  1.403713e+06
2017-07  1.375008e+06
2017-08  1.130442e+06
2017-09  8.703196e+05
2017-10  4.184447e+05
2017-11  4.406235e+04
2017-12  3.224408e+04

Let's seperate our data by segments using `index`:

In [50]:
prism = cl.Triangle(
    data=prism_df,
    origin="AccidentDate",
    origin_format="%Y-%m-%d",
    development="PaymentDate",
    development_format="%Y-%m-%d",
    columns=["Paid", "Incurred"],
    index="Line",
    cumulative=False,
)
prism

Triangle Summary
Valuation:           2017-12
Grain:                  OMDM
Shape:      (2, 2, 120, 120)
Index:                [Line]
Columns:    [Paid, Incurred]

We can futher `index` by coverages, or sublines:

In [51]:
prism = cl.Triangle(
    data=prism_df,
    origin="AccidentDate",
    origin_format="%Y-%m-%d",
    development="PaymentDate",
    development_format="%Y-%m-%d",
    columns=["Paid", "Incurred"],
    index=["Line", "Type"],
    cumulative=False,
)
prism

Triangle Summary
Valuation:           2017-12
Grain:                  OMDM
Shape:      (2, 2, 120, 120)
Index:          [Line, Type]
Columns:    [Paid, Incurred]

## Triangle Methods not Available in Pandas
Up until now, we've kept pretty close to the pandas API for triangle manipulation. However, there are data transformations commonly applied to triangles that don't have a nice `pandas` analogy.

For example, we often want to convert a triangle from an incremental view into a cumulative view and vice versa. This can be accomplished with the `incr_to_cum` and `cum_to_incr` methods.

In [52]:
prism["Paid"].sum()

1             2              3              4              5              6              7              8              9              10             11             12             13             14             15             16             17             18             19             20             21             22             23             24             25             26            27            28            29            30            31            32            33            34            35            36            37             38             39             40             41            42            43            44            45            46            47            48            49            50            51            52            53            54            55            56            57            58            59            60            61            62            63            64           65            66            67           68            69   70            71            72          73           74       75            76           77   78           79           80   81          82          83      84   85      86            87   88   89           90           91   92   93           94   95   96   97           98   99   100  101  102  103  104  105     106  107  108  109  110  111  112  113  114  115  116  117  118  119  120
2008-01           NaN           NaN   46915.015123   19898.778972   57215.983936   89783.258084   18301.827646   95054.432894   10332.363153   91772.229823   50403.256100   25904.392469   20363.039991   53973.827076   42845.725397   72887.347019   45247.565785   41837.701048   19000.000000   20681.849010    7437.519458   66622.733500  277594.082220   21437.716875   31795.146380  818091.638041  3.775590e+05  6.231213e+05  4.961923e+05  3.729639e+05  1.613041e+06  1.510690e+06  1.463860e+06  2.705437e+06  1.258599e+06  6.154958e+05  3.662470e+05  285689.926240  187000.000000            NaN    8733.822202           NaN   7324.582575           NaN           NaN           NaN           NaN           NaN           NaN   6908.747829           NaN           NaN           NaN           NaN           NaN           NaN           NaN   5659.301494           NaN           NaN   7509.548931           NaN           NaN           NaN          NaN           NaN           NaN          NaN           NaN  NaN           NaN           NaN         NaN          NaN      NaN           NaN          NaN  NaN          NaN          NaN  NaN         NaN         NaN     NaN  NaN     NaN           NaN  NaN  NaN          NaN          NaN  NaN  NaN          NaN  NaN  NaN  NaN          NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2008-02           NaN  28749.365091   22109.176517   79032.630204   63454.689224   59993.109001   64682.696457   68501.669363   33694.528307   71670.163132   10845.932008   76992.306674   45118.695949   31260.402970   31628.854136   44746.068059    5437.649534   28559.099760   20605.070884   23776.507583  243735.616281  127662.219754   63657.136032   23716.146264  341144.987773  963831.511040  1.269554e+06  7.748447e+05  9.630401e+05  1.733778e+06  1.207842e+06  1.042254e+06  1.637606e+06  7.552154e+05  8.835821e+05  8.399512e+05  1.800000e+05  180000.000000  360000.000000            NaN   14000.000000           NaN           NaN           NaN   7000.000000           NaN  13471.002738   9292.483535           NaN   2696.680976           NaN           NaN           NaN   4607.561736  19000.000000           NaN  29395.358080           NaN           NaN           NaN           NaN           NaN           NaN           NaN          NaN           NaN           NaN  5688.953874           NaN  NaN   6803.309901           NaN         NaN          NaN      NaN           NaN          NaN  NaN          NaN          NaN  NaN         NaN         NaN     NaN  NaN     NaN           NaN  NaN  NaN          NaN          NaN  NaN  NaN      

In [53]:
prism_cum = prism.incr_to_cum()
prism_cum["Paid"].sum()

1              2              3              4              5              6              7              8              9             10            11            12            13            14            15            16            17            18            19            20            21            22            23            24            25            26            27            28            29            30            31            32            33            34            35            36            37            38            39            40            41            42            43            44            45            46            47            48            49            50            51            52            53            54            55            56            57            58            59            60            61            62            63            64            65            66            67            68            69            70            71            72            73            74            75            76            77            78            79            80            81            82            83            84            85            86            87            88            89            90            91            92            93            94            95            96            97            98            99            100           101           102           103           104           105           106           107           108           109           110           111           112           113           114           115           116           117           118           119           120
2008-01           NaN            NaN   46915.015123   66813.794095  124029.778031  213813.036115  232114.863761  327169.296655  337501.659808  4.292739e+05  4.796771e+05  5.055815e+05  5.259446e+05  5.799184e+05  6.227641e+05  6.956515e+05  7.408990e+05  7.827367e+05  8.017367e+05  8.224186e+05  8.298561e+05  8.964788e+05  1.174073e+06  1.195511e+06  1.227306e+06  2.045397e+06  2.422956e+06  3.046078e+06  3.542270e+06  3.915234e+06  5.528275e+06  7.038966e+06  8.502826e+06  1.120826e+07  1.246686e+07  1.308236e+07  1.344861e+07  1.373430e+07  1.392130e+07  1.392130e+07  1.393003e+07  1.393003e+07  1.393735e+07  1.393735e+07  1.393735e+07  1.393735e+07  1.393735e+07  1.393735e+07  1.393735e+07  1.394426e+07  1.394426e+07  1.394426e+07  1.394426e+07  1.394426e+07  1.394426e+07  1.394426e+07  1.394426e+07  1.394992e+07  1.394992e+07  1.394992e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07  1.395743e+07
2008-02           NaN   28749.365091   50858.541608  129891.171812  193345.861036  253338.970037  318021.666494  386523.335857  420217.864164  4.918880e+05  5.027340e+05  5.797263e+05  6.248450e+05  6.561054e+05  6.877342e+05  7.324803e+05  7.379179e+05  7.664770e+05  7.870821e+05  8.108586e+05  1.054594e+06  1.182256e+06  1.245914e+06  1.269630e+06  1.610775e+06  2.574606e+06  3.844160e+06  4.619005e+06  5.582045e+06  7.315823e+06  8.523665e+06  9.565918e+06  1.120352e+07  1.195874e+07  1.284232e+07  1.368227e+07  1.386227e+07  1.404227e+07  1.440227e+07  1.440227e+07  1.441627e+07  1.441627e+07  1.441627e+07  1.44162

In [54]:
prism_incr = prism_cum.cum_to_incr()
prism_incr["Paid"].sum()

1             2              3              4              5              6              7              8              9              10             11             12             13             14             15             16             17             18             19             20             21             22             23             24             25             26            27            28            29            30            31            32            33            34            35            36            37             38             39             40             41            42            43            44            45            46            47            48            49            50            51            52            53            54            55            56            57            58            59            60            61            62            63            64           65            66            67           68            69   70            71            72          73           74       75            76           77   78           79           80   81          82          83      84   85      86            87   88   89           90           91   92   93           94   95   96   97           98   99   100  101  102  103  104  105     106  107  108  109  110  111  112  113  114  115  116  117  118  119  120
2008-01           NaN           NaN   46915.015123   19898.778972   57215.983936   89783.258084   18301.827646   95054.432894   10332.363153   91772.229823   50403.256100   25904.392469   20363.039991   53973.827076   42845.725397   72887.347019   45247.565785   41837.701048   19000.000000   20681.849010    7437.519458   66622.733500  277594.082220   21437.716875   31795.146380  818091.638041  3.775590e+05  6.231213e+05  4.961923e+05  3.729639e+05  1.613041e+06  1.510690e+06  1.463860e+06  2.705437e+06  1.258599e+06  6.154958e+05  3.662470e+05  285689.926240  187000.000000            NaN    8733.822202           NaN   7324.582575           NaN           NaN           NaN           NaN           NaN           NaN   6908.747829           NaN           NaN           NaN           NaN           NaN           NaN           NaN   5659.301494           NaN           NaN   7509.548931           NaN           NaN           NaN          NaN           NaN           NaN          NaN           NaN  NaN           NaN           NaN         NaN          NaN      NaN           NaN          NaN  NaN          NaN          NaN  NaN         NaN         NaN     NaN  NaN     NaN           NaN  NaN  NaN          NaN          NaN  NaN  NaN          NaN  NaN  NaN  NaN          NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2008-02           NaN  28749.365091   22109.176517   79032.630204   63454.689224   59993.109001   64682.696457   68501.669363   33694.528307   71670.163132   10845.932008   76992.306674   45118.695949   31260.402970   31628.854136   44746.068059    5437.649534   28559.099760   20605.070884   23776.507583  243735.616281  127662.219754   63657.136032   23716.146264  341144.987773  963831.511040  1.269554e+06  7.748447e+05  9.630401e+05  1.733778e+06  1.207842e+06  1.042254e+06  1.637606e+06  7.552154e+05  8.835821e+05  8.399512e+05  1.800000e+05  180000.000000  360000.000000            NaN   14000.000000           NaN           NaN           NaN   7000.000000           NaN  13471.002738   9292.483535           NaN   2696.680976           NaN           NaN           NaN   4607.561736  19000.000000           NaN  29395.358080           NaN           NaN           NaN           NaN           NaN           NaN           NaN          NaN           NaN           NaN  5688.953874           NaN  NaN   6803.309901           NaN         NaN          NaN      NaN           NaN          NaN  NaN          NaN          NaN  NaN         NaN         NaN     NaN  NaN     NaN           NaN  NaN  NaN          NaN          NaN  NaN  NaN      

By default (and in concert with the `pandas` philosophy), the methods associated with the `Triangle` class strive for immutability. This means that the `incr_to_cum` and `cum_to_incr` methods will return new a new triangle object that must be assigned, or it is thrown away. Many of the `chainladder.Triangle` methods have an `inplace` argument, or alternatively you can just use variable reassignment to store the transformed triangle object.

In [55]:
# This works
prism.incr_to_cum(inplace=True)
# So does this
prism = prism.incr_to_cum()

When dealing with triangles that have an `origin` axis, `development` axis, or both at a monthly or quarterly grain, the triangle can be summarized to a higher grain using the `grain` method.

The grain to which you want your triangle converted, specified as "OxDy" where "x" and "y" can take on values of "M", "Q", or "Y". For example:
* `grain(OYDY)` for Origin Year x Development Year.
* `grain(OQDM)` for Origin Quarter x Development Month.

In [56]:
prism_OYDY = prism.grain("OYDY")
prism_OYDY["Paid"].sum()

12            24            36            48            60            72            84            96            108           120
2008  3.404254e+06  1.119109e+07  7.461301e+07  1.503428e+08  1.509829e+08  1.511527e+08  1.512289e+08  1.512648e+08  1.512772e+08  1.512842e+08
2009  3.609385e+06  1.100293e+07  8.072635e+07  1.569708e+08  1.575995e+08  1.576971e+08  1.577364e+08  1.577434e+08  1.577487e+08           NaN
2010  4.067321e+06  1.239678e+07  7.421004e+07  1.610496e+08  1.616415e+08  1.617871e+08  1.618596e+08  1.618702e+08           NaN           NaN
2011  4.125232e+06  1.318314e+07  8.123977e+07  1.614129e+08  1.621876e+08  1.624175e+08  1.624907e+08           NaN           NaN           NaN
2012  4.584036e+06  1.400118e+07  7.779452e+07  1.521184e+08  1.525881e+08  1.528195e+08           NaN           NaN           NaN           NaN
2013  4.889623e+06  1.460774e+07  8.441850e+07  1.611103e+08  1.616730e+08           NaN           NaN           NaN           NaN           NaN
2014  5.546158e+06  1.640813e+07  7.725679e+07  1.549699e+08           NaN           NaN           NaN           NaN           NaN           NaN
2015  5.909029e+06  1.742761e+07  8.091458e+07           NaN           NaN           NaN           NaN           NaN           NaN           NaN
2016  6.080962e+06  1.858806e+07           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
2017  6.396536e+06           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN

Depending on the type of analysis being done, it may be more convenient to look at a triangle with its `development` axis expressed as a valuation rather than an age. This is also what the Schedule Ps look like. To do this, `Triangle` has two methods for toggling between a development triangle and a valuation triangle. The methods are `dev_to_val` and its inverse `val_to_dev`.

In [57]:
prism_OYDY_val = prism_OYDY.dev_to_val()
prism_OYDY_val["Paid"].sum()

2008          2009          2010          2011          2012          2013          2014          2015          2016          2017
2008  3.404254e+06  1.119109e+07  7.461301e+07  1.503428e+08  1.509829e+08  1.511527e+08  1.512289e+08  1.512648e+08  1.512772e+08  1.512842e+08
2009           NaN  3.609385e+06  1.100293e+07  8.072635e+07  1.569708e+08  1.575995e+08  1.576971e+08  1.577364e+08  1.577434e+08  1.577487e+08
2010           NaN           NaN  4.067321e+06  1.239678e+07  7.421004e+07  1.610496e+08  1.616415e+08  1.617871e+08  1.618596e+08  1.618702e+08
2011           NaN           NaN           NaN  4.125232e+06  1.318314e+07  8.123977e+07  1.614129e+08  1.621876e+08  1.624175e+08  1.624907e+08
2012           NaN           NaN           NaN           NaN  4.584036e+06  1.400118e+07  7.779452e+07  1.521184e+08  1.525881e+08  1.528195e+08
2013           NaN           NaN           NaN           NaN           NaN  4.889623e+06  1.460774e+07  8.441850e+07  1.611103e+08  1.616730e+08
2014           NaN           NaN           NaN           NaN           NaN           NaN  5.546158e+06  1.640813e+07  7.725679e+07  1.549699e+08
2015           NaN           NaN           NaN           NaN           NaN           NaN           NaN  5.909029e+06  1.742761e+07  8.091458e+07
2016           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  6.080962e+06  1.858806e+07
2017           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  6.396536e+06

When working with real-world data, the triangles can have holes, such as missing evluation(s), or no losses in certain origin(s). In these cases, it doesn't make sense to include empty accident periods or development ages. For example, in the `prism` dataset, the "Home" line has its latest accidents through 2016, and have no payments in development age '12'. Sometimes, dropping the non-applicable fields is usefule with the `dropna()` method.

In [58]:
prism_OYDY.loc["Home"]["Paid"]

12            24            36            48            60            72            84            96            108           120
2008  NaN  1.129305e+06  6.165887e+07  1.365206e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08
2009  NaN  1.872920e+05  6.713460e+07  1.422679e+08  1.425479e+08  1.425479e+08  1.425479e+08  1.425479e+08  1.425479e+08           NaN
2010  NaN  6.206026e+05  5.908246e+07  1.447572e+08  1.447572e+08  1.447572e+08  1.447572e+08  1.447572e+08           NaN           NaN
2011  NaN  5.032956e+05  6.504805e+07  1.439716e+08  1.442516e+08  1.442516e+08  1.442516e+08           NaN           NaN           NaN
2012  NaN  5.992768e+05  6.053664e+07  1.334124e+08  1.334124e+08  1.334124e+08           NaN           NaN           NaN           NaN
2013  NaN  5.363029e+05  6.644316e+07  1.416458e+08  1.416458e+08           NaN           NaN           NaN           NaN           NaN
2014  NaN  9.659729e+05  5.739426e+07  1.335428e+08           NaN           NaN           NaN           NaN           NaN           NaN
2015  NaN  3.710149e+05  5.904711e+07           NaN           NaN           NaN           NaN           NaN           NaN           NaN
2016  NaN  6.401785e+05           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
2017  NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN

Let's see what happens if we have no data for 2011.

In [59]:
prism_df_2011 = prism_df.copy()
prism_df_2011.loc[
    (prism_df_2011["AccidentDate"] >= "2011-01-01")
    & (prism_df_2011["AccidentDate"] < "2012-01-01"),
    "Paid",
] = None  # Let's assume we hav no payments for losses occurred in 2011
prism_2011 = cl.Triangle(
    data=prism_df_2011,
    origin="AccidentDate",
    origin_format="%Y-%m-%d",
    development="PaymentDate",
    development_format="%Y-%m-%d",
    columns=["Paid", "Incurred"],
    index=["Line", "Type"],
    cumulative=False,
)
prism_2011.incr_to_cum(inplace=True)
prism_2011_OYDY = prism_2011.grain("OYDY")
prism_2011_OYDY.loc["Home"]["Paid"]

12            24            36            48            60            72            84            96            108           120
2008  NaN  1.129305e+06  6.165887e+07  1.365206e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08
2009  NaN  1.872920e+05  6.713460e+07  1.422679e+08  1.425479e+08  1.425479e+08  1.425479e+08  1.425479e+08  1.425479e+08           NaN
2010  NaN  6.206026e+05  5.908246e+07  1.447572e+08  1.447572e+08  1.447572e+08  1.447572e+08  1.447572e+08           NaN           NaN
2011  NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
2012  NaN  5.992768e+05  6.053664e+07  1.334124e+08  1.334124e+08  1.334124e+08           NaN           NaN           NaN           NaN
2013  NaN  5.363029e+05  6.644316e+07  1.416458e+08  1.416458e+08           NaN           NaN           NaN           NaN           NaN
2014  NaN  9.659729e+05  5.739426e+07  1.335428e+08           NaN           NaN           NaN           NaN           NaN           NaN
2015  NaN  3.710149e+05  5.904711e+07           NaN           NaN           NaN           NaN           NaN           NaN           NaN
2016  NaN  6.401785e+05           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
2017  NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN

Note that the `dropna()` method will retain empty periods if they are surrounded by non-empty periods with valid data.

In [60]:
prism_2011_OYDY_droppedna = prism_2011_OYDY.loc["Home"].dropna()
prism_2011_OYDY_droppedna.loc["Home"]["Paid"]

24            36            48            60            72            84            96            108           120
2008  1.129305e+06  6.165887e+07  1.365206e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08
2009  1.872920e+05  6.713460e+07  1.422679e+08  1.425479e+08  1.425479e+08  1.425479e+08  1.425479e+08  1.425479e+08           NaN
2010  6.206026e+05  5.908246e+07  1.447572e+08  1.447572e+08  1.447572e+08  1.447572e+08  1.447572e+08           NaN           NaN
2011           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
2012  5.992768e+05  6.053664e+07  1.334124e+08  1.334124e+08  1.334124e+08           NaN           NaN           NaN           NaN
2013  5.363029e+05  6.644316e+07  1.416458e+08  1.416458e+08           NaN           NaN           NaN           NaN           NaN
2014  9.659729e+05  5.739426e+07  1.335428e+08           NaN           NaN           NaN           NaN           NaN           NaN
2015  3.710149e+05  5.904711e+07           NaN           NaN           NaN           NaN           NaN           NaN           NaN
2016  6.401785e+05           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN

## Commutative Properties of Triangle Methods
Where it makes sense, which is in most cases, the methods described above are commutative and can be applied in any order.

In [61]:
print("Commutative?", prism.sum().latest_diagonal == prism.latest_diagonal.sum())
print("Commutative?", prism.loc["Home"].link_ratio == prism.link_ratio.loc["Home"])

Commutative? True
Commutative? True


## Triangle Imports and Exports
To the extent the `Triangle` can be expressed as a `pandas.DataFrame`, you can use any of the pandas IO to send the data in and out. Note that converting to pandas is a one-way ticket with no inverse functions.

Another useful function is to copy the triangle and put it in the clipboard so we can paste it elsewhere (such as Excel):

In [62]:
prism_OYDY.loc["Home", "Paid"]

12            24            36            48            60            72            84            96            108           120
2008  NaN  1.129305e+06  6.165887e+07  1.365206e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08  1.368006e+08
2009  NaN  1.872920e+05  6.713460e+07  1.422679e+08  1.425479e+08  1.425479e+08  1.425479e+08  1.425479e+08  1.425479e+08           NaN
2010  NaN  6.206026e+05  5.908246e+07  1.447572e+08  1.447572e+08  1.447572e+08  1.447572e+08  1.447572e+08           NaN           NaN
2011  NaN  5.032956e+05  6.504805e+07  1.439716e+08  1.442516e+08  1.442516e+08  1.442516e+08           NaN           NaN           NaN
2012  NaN  5.992768e+05  6.053664e+07  1.334124e+08  1.334124e+08  1.334124e+08           NaN           NaN           NaN           NaN
2013  NaN  5.363029e+05  6.644316e+07  1.416458e+08  1.416458e+08           NaN           NaN           NaN           NaN           NaN
2014  NaN  9.659729e+05  5.739426e+07  1.335428e+08           NaN           NaN           NaN           NaN           NaN           NaN
2015  NaN  3.710149e+05  5.904711e+07           NaN           NaN           NaN           NaN           NaN           NaN           NaN
2016  NaN  6.401785e+05           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
2017  NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN

```python
prism_OYDY.loc["Home", "Paid"].to_clipboard()
```

Try to paste it elsewhere.

Alternatively, if you want to store the triangle elsewhere but be able to reconstitute a triangle out of it later, then you can use:
* `Triangle.to_json` and its inverse `cl.read_json` for json format<br>
* `Triangle.to_pickle` and its inverse `cl.read_pickle` for pickle format<br>

These have the added benefit of working on multi-dimensional triangles that don't fit into a DataFrame.

## Exercises

In [63]:
prism = cl.Triangle(
    data=prism_df,
    origin="AccidentDate",
    development="PaymentDate",
    columns=["Paid", "Incurred"],
    index=["Line", "Type"],  # multiple indices
    cumulative=False,
).incr_to_cum()
prism

Triangle Summary
Valuation:           2017-12
Grain:                  OMDM
Shape:      (2, 2, 120, 120)
Index:          [Line, Type]
Columns:    [Paid, Incurred]

1. What is the case incurred activity for calendar periods in 2015Q2 (March, April, and May in 2015) by "Line"?

In [64]:
incr_by_line = prism.groupby("Line").sum().cum_to_incr()["Incurred"].dev_to_val()
incr_by_line[
    (incr_by_line.valuation >= "2015-04-01") & (incr_by_line.valuation < "2015-07-01")
].sum("origin").to_frame(origin_as_datetime=True).astype(int)

development   2015-04   2015-05   2015-06
Line                                     
Auto          1883976   1704338   1692254
Home         10367939  10813821  14537617

2. For accident year 2015, what proportion of our Paid amounts come from each "Type" of claims?

In [65]:
prism_OYDY = prism.grain("OYDY")
prism_OYDY

Triangle Summary
Valuation:           2017-12
Grain:                  OYDY
Shape:        (2, 2, 10, 10)
Index:          [Line, Type]
Columns:    [Paid, Incurred]

In [66]:
by_type = (
    prism_OYDY.latest_diagonal[prism_OYDY.origin == "2015"]["Paid"]
    .groupby("Type")
    .sum()
    .to_frame(origin_as_datetime=True)
)
by_type / by_type.sum()

Type
Dwelling    0.729746
PD          0.270254
dtype: float64